src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp

```diff
@@ -558,6 +569,148 @@ namespace OpenMS
   }
 
 
+  void TheoreticalSpectrumGenerator::addInternalFragmentPeaks_(PeakSpectrum& spectrum,
+                                               const AASequence& peptide,
+                                               DataArrays::StringDataArray& ion_names,
+                                               DataArrays::IntegerDataArray& charges,
+                                               MSSpectrum::Chunks& chunks,
+                                               const Residue::ResidueType res_type,
+                                               Int charge) const
+  {
+    static double stat_a = Residue::getInternalToAIon().getMonoWeight();
+    static double stat_b = Residue::getInternalToBIon().getMonoWeight();
+
+    int f = 1 + int(add_isotopes_) + int(add_losses_); // TODO: calculate number of internal fragments
+    spectrum.reserve(spectrum.size() + f * peptide.size());
+
+    // precompute formula_str_cache
+    std::map<EmpiricalFormula, String> formula_str_cache;
+    if (add_losses_)
+    {
+      for (auto& p : peptide)
+      {
+        for (auto& formula : p.getLossFormulas())
+        {
+          String& loss_name = formula_str_cache[formula];
+          if (loss_name.empty())
+          {
+            loss_name = formula.toString();
```

Comment:
dont you want to populate the `formula_str_cache` at this point? or am I missing something?
```suggestion
            loss_name = formula.toString();
            formula_str_cache[formula] = loss_name;
```